### PR TITLE
Dracut initramfs boot

### LINKED ISSUE #1115: Add a bootstrap stage to allow a two stage node boot process
### Summary

Request is for `warewulfd` to serve a 

* stage=bootstrap_container which would download a container to be specified by `BootStrapContainer` in the node config
* stage=bootstrap_kernel which would download a kernel to be specified by `BootStrapKernel` in the node config
* stage=bootstrap_system which would download the system overlay for the `BootStrapSystemOverlay`
* stage=bootstrap_runtime which would download the runtime overlay for the `BootStrapRuntimeOverlay` 

This would allow optionally booting nodes with an arbitrary kernel and container, which could then perform whatever actions are needed before downloading the final kernel/container/overlays and pivot/switch-root-ing  and/or doing a `kexec()`. 

### Rationale

This enhancement would allow easily testing using a shim bootstrap, whether that is a dracut initrd, custom bootstrap image or just a minimal OS container.  Potential applications/benefits are:

* Making large containers possible with Legacy BIOS by loading them from a shim minimal OS
* Allowing for a full OS environment to set up disks, `grub` and any other steps required to statefully provision nodes
* Allow stateless provisioning to be direct to disk rather than using a `tmpfs` and depending on swapping to free memory

### Description

* Add methods to set the values of `BootStrap [Container|Kernel|SystemOverlay|RuntimeOverlay]` in the node config
* Add necessary `stage_*` requests to `warewulfd`

The BootStrap containers, kernels and overlays can be managed just like any other containers, kernels and overlays. 

### Additional information

Not sure how this would look for `ipxe` or `grub` or any other boot methods, so leaving those as an exercise for the admin after the bootstrrap functionality is available. 

### General information

- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by anderbubble**
## Description of the Pull Request (PR):

Support booting a dracut image from a container and using it to bootstrap a large container.

### To-do

- [ ] GRUB boot support
- [ ] Kernel override support

## This fixes or addresses the following GitHub issues:

 - Closes #1115

## Before submitting a PR, make sure you have done the following:

- [x] Signed off on all commits (e.g., using `git commit --signoff`) in agreement to the [DCO](DCO.txt)
- [x] Read the [documentation for contributing to Warewulf](https://warewulf.org/docs/main/contributing/contributing.html)
- [x] Added changes to the [CHANGELOG](https://github.com/warewulf/warewulf/blob/main/CHANGELOG.md) if necessary
- [x] Updated [userdocs](https://github.com/warewulf/warewulf/tree/main/userdocs) if necessary
- [x] Based this PR against the appropriate branch (typically [main](https://github.com/warewulf/warewulf/tree/main/userdocs))
- [x] Added myself as a contributor to the [Contributors File](https://github.com/warewulf/warewulf/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

**Comment by anderbubble** [comment]
@JasonYangShadow I refactored our work here, squashing redundant commits and making sure everything had a clear progression and commit messages. Let me know what you think.

File: internal/pkg/warewulfd/provision.go [hunk 1]
```diff
@@ -103,7 +107,8 @@ func ProvisionSend(w http.ResponseWriter, req *http.Request) {
 			Hwaddr:         rinfo.hwaddr,
 			ContainerName:  node.ContainerName.Get(),
 			KernelArgs:     node.Kernel.Args.Get(),
-			KernelOverride: node.Kernel.Override.Get()}
+			KernelOverride: node.Kernel.Override.Get(),
+			NetDevs:        tstruct.NetDevs}
```
Comment (anderbubble) [inline internal/pkg/warewulfd/provision.go:111]: @JasonYangShadow can I get you to add a test for my changes here?

File: internal/pkg/container/config.go [hunk 2]
```diff
@@ -27,3 +28,19 @@ func ImageParentDir() string {
...
+		return image_path
+	}
+
+	image_path = path.Join(RootFsDir(image), "boot", "initrd-"+kver)
+	if util.IsFile(image_path) {
+		return image_path
+	}
+
+	return ""
+}
```
Comment (anderbubble) [inline internal/pkg/container/config.go:46]: @JasonYangShadow I had to add an additional search pattern for Suse. I'm not super happy with this implementation, so feel free to improve it.

I squashed this into your commit, so it's still attributed to you. Let me know if you have any concerns.

File: internal/pkg/warewulfd/provision.go (hunk 1 above)
Comment (JasonYangShadow) [reply internal/pkg/warewulfd/provision.go:111]: @anderbubble sure will do that

File: dracut/modules.d/90wwinit/parse-wwinit.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
```
Comment (JasonYangShadow) [inline dracut/modules.d/90wwinit/parse-wwinit.sh:1]: is it /bin/bash here?

File: internal/pkg/warewulfd/provision.go (hunk 1 above)
Comment (JasonYangShadow) [reply internal/pkg/warewulfd/provision.go:111]: I've pushed test cases in this commit
https://github.com/warewulf/warewulf/pull/1191/commits/75c70fe3e2893808f9bff97049a1cb1b434c55b2

File: internal/pkg/container/config.go (hunk 2 above)
Comment (JasonYangShadow) [reply internal/pkg/container/config.go]: I've made some changes in this commit
https://github.com/warewulf/warewulf/pull/1191/commits/75c70fe3e2893808f9bff97049a1cb1b434c55b2

File: internal/pkg/warewulfd/provision.go (hunk 1 above)
Comment (anderbubble) [reply internal/pkg/warewulfd/provision.go:111]: Thanks!

**Comment by anderbubble** [comment]
Rebased on main.

File: dracut/modules.d/90wwinit/parse-wwinit.sh (hunk 3 above)
Comment (anderbubble) [reply dracut/modules.d/90wwinit/parse-wwinit.sh:1]: The other modules I looked at use `/bin/sh`. I kept this same pattern thinking that a POSIX shell may be all that is available in the initramfs environment.

File: internal/pkg/container/config.go (hunk 2 above)
Comment (anderbubble) [reply internal/pkg/container/config.go]: @JasonYangShadow I changed it a bit against to more closely match what's in `internal/pkg/kernel/kernel.go`.

I also removed a test case that dependendent on the order of files being returned, as I'd consider that undefined. (I know the golang Walk makes it deterministic, but I'd rather that not be an assumption.

**Comment by anderbubble** [comment]
@mslacken @jcsiadal and @griznog I'd still like to add Grub support and see if I can get kernel overrides working for this; but I'd love to get your thoughts on getting this merged in time for v4.5.2. Let me know what you think.

Thanks!

**Comment by mslacken** [comment]
If we link this together with #815 we have state full install. Steps for this would be:
* add ignition to dracut
* fetch the igniton.json from the warewulf host, which was the purpose of #815
* make sure that $NEWROOT is correctly set

This would allow a dump of the container to the rootfs. What is till missing would be checks, if the rootfs needs to be updated or reinstalled. I think this could achieved by placing configurations in `/warewulf` on the node.

**Review by mslacken — APPROVED** [review]

**Comment by mslacken** [comment]
@anderbubble my commit is WIP, some args get lost

**Comment by mslacken** [comment]
@anderbubble @JasonYangShadow Buidling the dracut works now on `wwctl container build` feel free to test it.

**Comment by anderbubble** [comment]
@mslacken I'm torn on your contribution here. For example, I definitely appreciate this fix:

https://github.com/warewulf/warewulf/pull/1191/commits/9f1d378bf8a6a26a30bc914f6b711c466f7296ef#diff-56ccdd3cf2e2cb4ea3726e8587dbf75472b67d8ba3c25b47b3b7d9a80135ef3dL167-R174

But, ultimately, I do not think that `container build` should presumptively build a dracut image just because it's installed. More likely I think dracut should get called by `/etc/warewulf/container_exit.sh`.

For now, I want to omit this change from the PR; but I'll try to make sure it doesn't get lost so that we can still discuss it.

**Comment by anderbubble** [comment]
> I'll try to make sure it doesn't get lost so that we can still discuss it.

Captured at https://github.com/anderbubble/warewulf/tree/dracut-initramfs-autobuild

**Pull request merged**